src/output-json.c

```diff
@@ -707,6 +710,76 @@ void CreateJSONFlowId(json_t *js, const Flow *f)
     }
 }
 
+static inline void JSONFormatAndAddMACAddr(json_t *parent, const char *key,
+                                   uint8_t *val, bool is_array)
+{
+    char eth_addr[19];
+    (void) snprintf(eth_addr, 19, "%02x:%02x:%02x:%02x:%02x:%02x",
+                    val[0], val[1], val[2], val[3], val[4], val[5]);
+    if (is_array) {
+        json_array_append_new(parent, json_string(eth_addr));
+    } else {
+        json_object_set_new(parent, key, json_string(eth_addr));
+    }
+}
+
+/* only required to traverse the MAC address set */
+typedef struct JSONMACAddrInfo {
+    json_t *src, *dst;
+} JSONMACAddrInfo;
+
+static int JSONFlowAppendMACAddrs(uint8_t *val, MacSetSide side, void *data)
+{
+    JSONMACAddrInfo *info = (JSONMACAddrInfo*) data;
+    if (side == MAC_SET_DST) {
+        JSONFormatAndAddMACAddr(info->dst, NULL, val, true);
+    } else {
+        JSONFormatAndAddMACAddr(info->src, NULL, val, true);
+    }
+    return 0;
+}
+
+int CreateJSONEther(json_t *parent, const Packet *p, const Flow *f)
+{
+    json_t *ejs = json_object();
+    if (unlikely(ejs == NULL))
+        return 0;
+    if (p == NULL) {
+        /* we are creating an ether object in a flow context, so we need to
+           append to arrays */
+        if (f != NULL && f->macset && MacSetSize(f->macset) > 0) {
+            JSONMACAddrInfo info;
+            info.dst = json_array();
+            info.src = json_array();
+            int ret = MacSetForEach(f->macset, JSONFlowAppendMACAddrs, &info);
+            if (unlikely(ret != 0)) {
+                /* should not happen, JSONFlowAppendMACAddrs is sane */
+                json_decref(info.dst);
+                json_decref(info.src);
+                json_decref(ejs);
+                return ret;
+            }
+            json_object_set_new(ejs, "dest_macs", info.dst);
```

Comment:
Is the Ethertype available? For consistency, it'd be helpful to output that -- either as a list or a single value.

Reply:
Not stored so far due to complexity. This also brings the question whether e want to keep an association between the ethertype and to the MAC contexts that it was seen in or not?

TBH I'd rather ask: "Do we need the ethertype at all?" Let's say the next layer is a protocol that Suricata can decode, then it would be visible in the metadata added by the next layer (e.g. IPv4/v6, VLAN, etc.). If the next layer is a protocol that Suricata can not decode, then we would not have anything to add this JSON to -- IIRC there is nothing to trigger a flow-context JSON output except for a app-layer tx or `flow` event (which would need to be based on a decodable protocol first) or an alert (where `ip` is the most generic protocol that can be defined in rules).
I've just included the ethertype so I do not deviate from Eric's first version, but personally I don't see much value there. Any comments?

Reply:
I'm not sure how much value having the ethertype adds. Perhaps others have opinions?

Reply:
I'll look into handling the ethertype the same way as the MAC addresses in my next PR, for the sake of completeness.

Reply:
After taking a second look, this would mean duplicating quite a bit of logic for maintaining buffers for different types (`MacAddr`s and `uint16_t`s). Before implementing this, I would like to find out if it is even possible to have multiple different ethertypes in this context, i.e. would such packets even be appearing in the same flow?

Reply:
I don't think the effort is worth it.